### ISSUE TITLE
[9.5.6] Update items causes then to remain locked

### ISSUE THREAD

**Issue by SogurB** [glpi-project/glpi#9555]
**Description**

With item locks enabled, trying to update anything within the "List of items to lock" results in the object locked until the user forces the unlock.

**To reproduce**

1. Enable "Use locks"
2. Open a ticket
3. Add a followup to a ticket as a user with update rights

Another example:
1. Enable "Use locks"
2. Open a computer
3. Make a change to a field and hit save. 

**Logs**
Logs are empty. Fresh installation.

**Expected behavior**

I don't expect the program to aware me that is locked as I'm making changes to it.

**Comment by ppenin** [comment]
same issue for us since 9.5.6

**Comment by andrepetinga** [comment]
Same issue!!! Just updated the quality environment to 9.5.6...
I Unlock the ticket, make a small change like "Category", hit Save and Locked!! Unlock it again and change another item, like "Request source", hit Save and Locked again!

**Comment by ppenin** [comment]
We have disabled ticket locking while waiting for a fix to be released

**Comment by ppenin** [comment]
any fix to be released for this issue ?

**Comment by ppenin** [comment]
any fix to be released for this issue ?

**Comment by hgpit** [comment]
We just upgraded from 9.5.5 to 9.5.6 and are seeing the same problem.

**Comment by b-keller** [comment]
Hi,
Same here, update from 9.5.5 to 9.5.6 > ticket unlock not working when exit ticket...

**Comment by MrYsson** [comment]
The issue is different according to the browser. If it's Firefox, the request is not always sent before the page is left. When it's Chromium, the parameters are not set correctly and they aren't accessible through `$_POST`.
The solution I found is to create 2 JavaScript functions that depend on the browser. If the `User-Agent` contains `Chrom`, we assume it's based on Chromium, otherwise it's Firefox.

The solution for Firefox is to use the `fallback_request` by default
For Chromium, we change the `Content-Type` to `application/x-www-form-urlencoded; charset=UTF-8` and we create an `URLSearchParams` with the parameters we want and use it as the body.

So I changed the lines 331-365 of the file ``glpi/inc/objectlock.class.php`` into this code :
```php
         echo  Html::scriptBlock("$(function() {
            $(window).on('beforeunload', function() {
               var fallback_request = function() {
                  $.post({
                     url: '"  .  $CFG_GLPI['root_doc'] .  "/ajax/unlockobject.php',
                     async: false,
                     cache: false,
                     data: {
                        unlock: 1,
                        id: $id
                     },
                     dataType: 'json'
                  });
               };

               if (typeof window.fetch !== 'undefined') {".
                (!str_contains($_SERVER['HTTP_USER_AGENT'],  'Chrom') ?  "fallback_request();
                return;":
                "var paramsString = 'unlock=1&id=$id';
                var searchParams = new URLSearchParams(paramsString);
                  fetch('"  .  $CFG_GLPI['root_doc'] .  "/ajax/unlockobject.php', {
                     method: 'POST',
                     cache: 'no-cache',
                     headers: {
                        'Accept': 'application/json',
                        'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8',
                        'X-Glpi-Csrf-Token': getAjaxCsrfToken()
                     },
                     body: searchParams
                  }).catch(function(error) {
                     //fallback if fetch fails
                     fallback_request();
                  });")."
               } else {
                  //fallback for browsers with no fetch support
                  fallback_request();
               }
            });
         })");
```
This code is probably far from perfect, but it works.

**Comment by ppenin** [comment]
Thanks for the fix 4eb42d7 . It works fine on 9.5.6

**Comment by andrepetinga** [comment]
Yes. It seems to work fine.
Thanks!

**Comment by cedric-anne** [comment]
@MrYsson , @hgpit , @ppenin , @andrepetinga , @tomolimo , @SogurB , @b-keller ,

Can you validate the fix proposed in #10438 ?

Regards

**Comment by b-keller** [comment]
Hello to all,
I just tested this, and for me, it works in chromium and firefox.

but there is still a refresh bug on firefox, the "locked by" field is not empty when you go back to the ticket list.

**Comment by ppenin** [comment]
It works like a charm in 9.5.6 with Google Chrome  97.0.4692.99

**Comment by andrepetinga** [comment]
Hello @cedric-anne.
When I applied the fix [4eb42d7](https://github.com/glpi-project/glpi/commit/4eb42d7db9f4250c289cdfe01e2af4ff494d2e95) it worked well.
Now, if I apply the [#10438](https://github.com/glpi-project/glpi/pull/10438), meaning remove the previous change and change in another file, it stopped working again... Tested on Chrome and Edge
